Strange CI failure of `docker-py`:
```
2022-10-31T09:55:18.0008654Z =================================== FAILURES ===================================
2022-10-31T09:55:18.0015712Z ___________________ SwarmTest.test_init_swarm_with_ca_config ___________________
2022-10-31T09:55:18.0021705Z docker/api/client.py:268: in _raise_for_status
2022-10-31T09:55:18.0027651Z     response.raise_for_status()
2022-10-31T09:55:18.0038028Z /usr/local/lib/python3.7/site-packages/requests/models.py:953: in raise_for_status
2022-10-31T09:55:18.0038470Z     raise HTTPError(http_error_msg, response=self)
2022-10-31T09:55:18.0038993Z E   requests.exceptions.HTTPError: 503 Server Error: Service Unavailable for url: http+docker://localhost/v1.43/swarm/init
2022-10-31T09:55:18.0039363Z 
2022-10-31T09:55:18.0039562Z During handling of the above exception, another exception occurred:
2022-10-31T09:55:18.0039991Z tests/integration/api_swarm_test.py:89: in test_init_swarm_with_ca_config
2022-10-31T09:55:18.0040368Z     assert self.init_swarm(swarm_spec=spec)
2022-10-31T09:55:18.0040921Z tests/integration/base.py:127: in init_swarm
2022-10-31T09:55:18.0041267Z     return self._init_swarm(self.client, **kwargs)
2022-10-31T09:55:18.0041605Z tests/integration/base.py:94: in _init_swarm
2022-10-31T09:55:18.0042069Z     '127.0.0.1', listen_addr=helpers.swarm_listen_addr(), **kwargs
2022-10-31T09:55:18.0042426Z docker/utils/decorators.py:34: in wrapper
2022-10-31T09:55:18.0042732Z     return f(self, *args, **kwargs)
2022-10-31T09:55:18.0043013Z docker/api/swarm.py:170: in init_swarm
2022-10-31T09:55:18.0043333Z     return self._result(response, json=True)
2022-10-31T09:55:18.0043651Z docker/api/client.py:274: in _result
2022-10-31T09:55:18.0057337Z     self._raise_for_status(response)
2022-10-31T09:55:18.0057674Z docker/api/client.py:270: in _raise_for_status
2022-10-31T09:55:18.0058023Z     raise create_api_error_from_http_exception(e)
2022-10-31T09:55:18.0058399Z docker/errors.py:31: in create_api_error_from_http_exception
2022-10-31T09:55:18.0058779Z     raise cls(e, response=response, explanation=explanation)
2022-10-31T09:55:18.0059481Z E   docker.errors.APIError: 503 Server Error for http+docker://localhost/v1.43/swarm/init: Service Unavailable ("Swarm is encrypted and needs to be unlocked before it can be used. Please use "docker swarm unlock" to unlock it.")
2022-10-31T09:55:18.0060417Z ------- generated xml file: /src/bundles/test-docker-py/junit-report.xml -------
```

Seems unrelated. I haven't seen this failure in other PRs though. Will re-run this job.

--- a/distribution/errors.go
+++ b/distribution/errors.go
@@ -63,6 +63,19 @@ func (e notFoundError) Cause() error {
 	return e.cause
 }
 
+// unsupportedMediaTypeError is an error issued when attempted
+// to pull unsupported content.
+type unsupportedMediaTypeError struct {
+	MediaType string
+}
+
+func (e unsupportedMediaTypeError) InvalidParameter() {}
+
+// Error returns the error string for unsupportedMediaTypeError.
+func (e unsupportedMediaTypeError) Error() string {
+	return "unsupported media type " + e.MediaType
+}
+
 // translatePullError is used to convert an error from a registry pull
 // operation to an error representing the entire pull operation. Any error
 // information which is not used by the returned error gets output to
@@ -124,6 +137,8 @@ func continueOnError(err error, mirrorEndpoint bool) bool {
 		// Failures from a mirror endpoint should result in fallback to the
 		// canonical repo.
 		return mirrorEndpoint
+	case unsupportedMediaTypeError:
+		return false
 	case error:
 		return !strings.Contains(err.Error(), strings.ToLower(syscall.ESRCH.Error()))
 	}
@@ -153,7 +168,7 @@ func retryOnError(err error) error {
 			return xfer.DoNotRetry{Err: v.Err}
 		}
 		return retryOnError(v.Err)
-	case *client.UnexpectedHTTPResponseError:
+	case *client.UnexpectedHTTPResponseError, unsupportedMediaTypeError:
 		return xfer.DoNotRetry{Err: err}
 	case error:
 		if err == distribution.ErrBlobUnknown {

--- a/distribution/pull_v2.go
+++ b/distribution/pull_v2.go
@@ -7,6 +7,7 @@ import (
 	"io"
 	"os"
 	"runtime"
+	"strings"
 	"time"
 
 	"github.com/containerd/containerd/log"
@@ -606,11 +607,30 @@ func (p *puller) pullSchema1(ctx context.Context, ref reference.Reference, unver
 	return imageID, manifestDigest, nil
 }
 
+func checkSupportedMediaType(mediaType string) error {
+	supportedMediaTypes := []string{
+		"application/vnd.oci.image.",
+		"application/vnd.docker.",
+	}
+
+	lowerMt := strings.ToLower(mediaType)
+	for _, mt := range supportedMediaTypes {
+		if strings.HasPrefix(lowerMt, mt) {
+			return nil
+		}
+	}
+	return unsupportedMediaTypeError{MediaType: mediaType}
+}
+
 func (p *puller) pullSchema2Layers(ctx context.Context, target distribution.Descriptor, layers []distribution.Descriptor, platform *specs.Platform) (id digest.Digest, err error) {
 	if _, err := p.config.ImageStore.Get(ctx, target.Digest); err == nil {
 		// If the image already exists locally, no need to pull
 		// anything.
 		return target.Digest, nil
+	}
+
+	if err := checkSupportedMediaType(target.MediaType); err != nil {
+		return "", err
 	}
 
 	var descriptors []xfer.DownloadDescriptor
@@ -620,6 +640,9 @@ func (p *puller) pullSchema2Layers(ctx context.Context, target distribution.Desc
 	for _, d := range layers {
 		if err := d.Digest.Validate(); err != nil {
 			return "", errors.Wrapf(err, "could not validate layer digest %q", d.Digest)
+		}
+		if err := checkSupportedMediaType(d.MediaType); err != nil {
+			return "", err
 		}
 		layerDescriptor := &layerDescriptor{
 			digest:          d.Digest,

